orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/support/TargetReferenceLinearStageSupport.groovy

```diff
@@ -36,33 +40,55 @@ abstract class TargetReferenceLinearStageSupport implements StageDefinitionBuild
   DetermineTargetReferenceStage determineTargetReferenceStage
 
   @Override
-  def List<Stage> aroundStages(Stage parentStage) {
-    return composeTargets(parentStage)
+  void beforeStages(@Nonnull Stage parent, @Nonnull StageGraphBuilder graph) {
+    log.warn("TargetReferenceLinearStageSupport is deprecated (execution: ${parent.execution.id})")
+
+    parent.resolveStrategyParams()
+
+    if (targetReferenceSupport.isDynamicallyBound(parent) && !parent.parentStageId) {
+      Map injectedContext = new HashMap(parent.context)
+      injectedContext.regions = new ArrayList(parent.context.regions)
+      graph.add {
+        it.type = determineTargetReferenceStage.type
+        it.name = "determineTargetReferences"
+        it.context = injectedContext
+      }
+    }
+  }
+
+  @Override
+  void afterStages(@Nonnull Stage parent, @Nonnull StageGraphBuilder graph) {
+    parent.resolveStrategyParams()
+
+    List<Stage> stages = composeTargets(parent, graph)
+    for (int i = 1; i < stages.size(); i++) {
+      graph.connect(stages.get(i - 1), stages.get(i))
```

Comment:
where does it add `stages[0]`?

Reply:
Updated to check for `size() > 1` before hitting this loop.